src/protocol/pki.rs

```diff
@@ -0,0 +1,426 @@
+    }
+
+    /// This method checks if the status is `Accepted` or not. It is used only in tests
+    fn is_accepted(&self) -> bool {
+        matches!(
+            self.0,
+            pki_enrollment_info::PkiEnrollmentInfoStatus::Accepted { .. }
+        )
+    }
+
+    /// This method checks if the status is `Rejected` or not. It is used only in tests
+    fn is_rejected(&self) -> bool {
+        matches!(
+            self.0,
+            pki_enrollment_info::PkiEnrollmentInfoStatus::Rejected { .. }
+        )
+    }
+}
+
+#[pyclass]
+pub(crate) struct PkiEnrollmentListReq(pub pki_enrollment_list::Req);
+
+#[pymethods]
+impl PkiEnrollmentListReq {
+    #[new]
+    fn new() -> Self {
+        Self(pki_enrollment_list::Req)
+    }
+
+    fn dump<'py>(&self, py: Python<'py>) -> PyResult<&'py PyBytes> {
+        Ok(PyBytes::new(
+            py,
+            &self
+                .0
+                .clone()
+                .dump()
+                .map_err(|e| ProtocolError::new_err(format!("encoding error: {e}")))?,
+        ))
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub(crate) struct PkiEnrollmentListItem(pki_enrollment_list::PkiEnrollmentListItem);
+
+#[pymethods]
+impl PkiEnrollmentListItem {
+    #[new]
+    fn new(
+        enrollment_id: EnrollmentID,
+        submit_payload: Vec<u8>,
+        submit_payload_signature: Vec<u8>,
+        submitted_on: DateTime,
+        submitter_der_x509_certificate: Vec<u8>,
+    ) -> Self {
+        Self(pki_enrollment_list::PkiEnrollmentListItem {
+            enrollment_id: enrollment_id.0,
+            submit_payload,
+            submit_payload_signature,
+            submitted_on: submitted_on.0,
+            submitter_der_x509_certificate,
+        })
+    }
+
+    #[getter]
+    fn enrollment_id(&self) -> EnrollmentID {
+        EnrollmentID(self.0.enrollment_id)
+    }
+
+    #[getter]
+    fn submit_payload<'py>(&'py self, py: Python<'py>) -> &'py PyBytes {
+        PyBytes::new(py, &self.0.submit_payload)
+    }
+
+    #[getter]
+    fn submit_payload_signature<'py>(&'py self, py: Python<'py>) -> &'py PyBytes {
+        PyBytes::new(py, &self.0.submit_payload_signature)
+    }
+
+    #[getter]
+    fn submitted_on(&self) -> DateTime {
+        DateTime(self.0.submitted_on)
+    }
+
+    #[getter]
+    fn submitter_der_x509_certificate<'py>(&'py self, py: Python<'py>) -> &'py PyBytes {
+        PyBytes::new(py, &self.0.submitter_der_x509_certificate)
+    }
+}
+
+gen_rep!(
+    pki_enrollment_list,
+    PkiEnrollmentListRep,
+    { .. },
+    [NotAllowed, reason: Reason]
+);
+
+#[pyclass(extends = PkiEnrollmentListRep)]
+pub(crate) struct PkiEnrollmentListRepOk;
+
+#[pymethods]
+impl PkiEnrollmentListRepOk {
+    #[new]
+    fn new(enrollments: Vec<PkiEnrollmentListItem>) -> (Self, PkiEnrollmentListRep) {
+        (
+            Self,
+            PkiEnrollmentListRep(pki_enrollment_list::Rep::Ok {
+                enrollments: enrollments.iter().map(|e| e.0.clone()).collect(),
```

Comment:
`enrollments.into_iter().map(|e| e.0).collect()`, to avoid copy

Reply:
Done.